torch_glow/src/GlowIValue.cpp

```diff
@@ -0,0 +1,281 @@
+
+#define ExpectTag(EXPECTED_TAG)                                                \
+  RETURN_ERR_IF_NOT(tag_ == (EXPECTED_TAG),                                    \
+                    strFormat("Expected GlowIValue with tag %s but found %s",  \
+                              tagToStr((EXPECTED_TAG)), tagToStr(tag_)))
+
+llvm::Expected<Tensor *> GlowIValue::toTensor() {
+  ExpectTag(Tag::Tensor);
+  return payload_.asTensor;
+}
+
+llvm::Expected<double> GlowIValue::toDouble() {
+  ExpectTag(Tag::Double);
+  return payload_.asDouble;
+}
+
+llvm::Expected<int32_t> GlowIValue::toInt() {
+  ExpectTag(Tag::Int);
+  return payload_.asInt;
+}
+
+llvm::Expected<bool> GlowIValue::toBool() {
+  ExpectTag(Tag::Bool);
+  return payload_.asBool;
+}
+
+llvm::Expected<std::vector<int32_t> *> GlowIValue::toIntList() {
+  ExpectTag(Tag::IntList);
+  return payload_.asIntList;
+}
+
+llvm::Expected<std::vector<double> *> GlowIValue::toDoubleList() {
+  ExpectTag(Tag::DoubleList);
+  return payload_.asDoubleList;
+}
+
+llvm::Expected<std::vector<bool> *> GlowIValue::toBoolList() {
+  ExpectTag(Tag::BoolList);
+  return payload_.asBoolList;
+}
+
+llvm::Expected<std::vector<Tensor> *> GlowIValue::toTensorList() {
+  ExpectTag(Tag::TensorList);
+  return payload_.asTensorList;
+}
+#undef ExpectTag
+
+void GlowIValue::fromNone() {
+  reset();
+  tag_ = Tag::None;
+}
+
+void GlowIValue::fromTensor(Tensor tensor) {
+  reset();
+  tag_ = Tag::Tensor;
+  payload_.asTensor = new glow::Tensor(std::move(tensor));
+}
+
+void GlowIValue::fromDouble(double d) {
+  reset();
+  tag_ = Tag::Double;
+  payload_.asDouble = d;
+}
+
+void GlowIValue::fromInt(int32_t i) {
+  reset();
+  tag_ = Tag::Int;
+  payload_.asInt = i;
+}
+
+void GlowIValue::fromBool(bool b) {
+  reset();
+  tag_ = Tag::Bool;
+  payload_.asBool = b;
+}
+
+void GlowIValue::fromIntList(std::vector<int32_t> intList) {
+  reset();
+  tag_ = Tag::IntList;
+  payload_.asIntList = new std::vector<int32_t>;
+  std::swap(intList, *payload_.asIntList);
+}
+
+void GlowIValue::fromDoubleList(std::vector<double> doubleList) {
+  reset();
+  tag_ = Tag::DoubleList;
+  payload_.asDoubleList = new std::vector<double>;
+  std::swap(doubleList, *payload_.asDoubleList);
+}
+
+void GlowIValue::fromBoolList(std::vector<bool> boolList) {
+  reset();
+  tag_ = Tag::BoolList;
+  payload_.asBoolList = new std::vector<bool>;
+  std::swap(boolList, *payload_.asBoolList);
+}
+
+void GlowIValue::fromTensorList(std::vector<Tensor> tensorList) {
+  reset();
+  tag_ = Tag::TensorList;
+  payload_.asTensorList = new std::vector<Tensor>;
+  std::swap(tensorList, *payload_.asTensorList);
+}
+
+void GlowIValue::fromTuple(std::vector<GlowIValue> glowIValList) {
+  reset();
+  tag_ = Tag::Tuple;
+  payload_.asTuple = new std::vector<GlowIValue>;
+  std::swap(glowIValList, *payload_.asTuple);
+}
+
+llvm::Error GlowIValue::fromIValue(const at::IValue &ival) {
+  reset();
+  if (ival.isNone()) {
```

Comment:
PyTorch does not provide something like `getTag()` so that one could write a switch statement?

Reply:
Unfortunately they do not, see this comment on why
https://github.com/pytorch/pytorch/blob/f8852c947bf527a994002aa04f8e77b78ac75f01/aten/src/ATen/core/ivalue.h#L435-L440